sklearn/tree/_utils.pyx

```diff
@@ -65,396 +66,325 @@ cdef inline float64_t rand_uniform(float64_t low, float64_t high,
-            self.sum_w_0_k -= weight
-
-            # minimize k such that sum(W[0:k]) >= total_weight / 2
-            # by incrementing k and updating sum_w_0_k accordingly
-            # until the condition is met.
-            while(self.k < self.samples.size() and
-                  (self.sum_w_0_k < self.total_weight / 2.0)):
-                self.k += 1
-                self.sum_w_0_k += self.samples.get_weight_from_index(self.k-1)
-            return 0
-
-        if data >= original_median:
-            # removing above the median
-            # minimize k such that sum(W[0:k]) >= total_weight / 2
-            while(self.k > 1 and ((self.sum_w_0_k -
-                                   self.samples.get_weight_from_index(self.k-1))
-                                  >= self.total_weight / 2.0)):
-                self.k -= 1
-                self.sum_w_0_k -= self.samples.get_weight_from_index(self.k)
-            return 0
-
-    cdef float64_t get_median(self) noexcept nogil:
-        """Write the median to a pointer, taking into account
-        sample weights."""
-        if self.sum_w_0_k == (self.total_weight / 2.0):
-            # split median
-            return (self.samples.get_value_from_index(self.k) +
-                    self.samples.get_value_from_index(self.k-1)) / 2.0
-        if self.sum_w_0_k > (self.total_weight / 2.0):
-            # whole median
-            return self.samples.get_value_from_index(self.k-1)
+    cdef inline void _heapify_down(self, intp_t i) noexcept nogil:
+        """Move down the element at index i until heap invariant is restored."""
+        cdef intp_t n = self.size
+        cdef intp_t left, right, mc
+        while True:
+            left = (i << 1) + 1
+            right = left + 1
+            if left >= n:
+                return
+            mc = left
+            if right < n and self.heap[right] < self.heap[left]:
+                mc = right
+            if self.heap[i] > self.heap[mc]:
+                self._swap(i, mc)
+                i = mc
+            else:
+                return
+
+    # Wrappers callable from Python for tests:
+
+    def _py_push(self, double value, double weight):
+        self.push(value, weight)
+
+    def _py_pop(self):
+        cdef double v, w
+        self.pop(&v, &w)
+        return v, w
+
+    def _py_reset(self):
+        self.reset()
 
+# =============================================================================
+# MAE split precomputations algorithm
+# =============================================================================
 
-def _any_isnan_axis0(const float32_t[:, :] X):
-    """Same as np.any(np.isnan(X), axis=0)"""
+cdef void precompute_absolute_errors(
+    const float64_t[:, ::1] ys,
+    const float64_t[:] sample_weight,
+    const intp_t[:] sample_indices,
+    WeightedHeap above,
+    WeightedHeap below,
+    intp_t k,
+    intp_t start,
+    intp_t end,
+    float64_t[::1] abs_errors,
+    float64_t[::1] medians
+) noexcept nogil:
```

Comment:
We should add a description of each argument like typical Python numpy-style docstring